feat: Add go router example

mbfakourii opened the pull request:
Add an example for the `go_router` package, based on the `go_router` [stateful_shell_route](https://github.com/flutter/packages/blob/main/packages/go_router/example/lib/stateful_shell_route.dart) example.

mbfakourii committed:
feat: Add go router example

c0pp1ce commented on example/lib/go_router_example.dart:
```diff
@@ -0,0 +1,306 @@
+      );
+}
+
+class ScaffoldWithNavBar extends StatelessWidget {
```
The name is quite misleading. It should better reflect that this is where the package comes into play. Something like ``AdaptiveNavigationShell`` would come to my mind.

mbfakourii replied:
Done

c0pp1ce commented:
I think it would also be good to add some comments at the relevant parts so that a reader can faster identify the important steps.

mbfakourii committed:
fix: changes and add document

Change `ScaffoldWithNavBar` to `AdaptiveNavigationShell`.
Add document and comments.

mbfakourii commented:
> I think it would also be good to add some comments at the relevant parts so that a reader can faster identify the important steps.

Done, may check again

c0pp1ce commented on example/lib/go_router_example.dart:
```diff
@@ -0,0 +1,345 @@
+        branches: <StatefulShellBranch>[
+          // The route branch for the first tab of the bottom navigation bar.
+          StatefulShellBranch(
+            navigatorKey: _sectionANavigatorKey,
```
Is this key actually used or can it be removed?

mbfakourii replied:
removed

c0pp1ce commented:
In a lot of comments you are still referring to bottom navigation bar. I would change those to something like "navigation shell" where applicable.

c0pp1ce commented on example/lib/go_router_example.dart:
```diff
@@ -0,0 +1,345 @@
+/// Builds the "shell" for the app by building a Scaffold with a
+/// BottomNavigationBar, where [child] is placed in the body of the Scaffold.
+class AdaptiveNavigationShell extends StatelessWidget {
+  /// Constructs an [AdaptiveNavigationShell].
```
This comment is redundant I think.

mbfakourii replied:
removed

c0pp1ce commented on example/lib/go_router_example.dart:
```diff
@@ -0,0 +1,345 @@
+          final int index,
+        ) =>
+            navigationShell.goBranch(index),
+        initialIndex: navigationShell.currentIndex,
```
This could be moved to above the ``onDestinationChanged`` so that it is not separating the navigation-related methods.

mbfakourii replied:
done

mbfakourii commented:
> In a lot of comments you are still referring to bottom navigation bar. I would change those to something like "navigation shell" where applicable.

Done, may check again

mbfakourii committed:
doc: Improve

mbfakourii commented:
cc @c0pp1ce

c0pp1ce commented:
I am very sorry for the late response. I have been quite busy (university, work).

I will look into this PR again once I examined if the example displays the "easiest" (as in: easy to adapt to more complex use cases) version of the go router integration.
